### ISSUE TITLE
App crashes on clicking download in release + minified buiilds

### ISSUE THREAD

**Issue by build3r** [microsoft/appcenter-sdk-android#1523]
### **Description**
The app crashes consistently when the user goes to background just before the apk download starts. There is no way from recovering this error when the update is mandatory.
The crash has occurred 684 times on the app affecting more than 200 users

Please describe the issue you are facing using the SDK.

### **Repro Steps**
1. Install old version with appcenter sdk v 4.1.1, with following build config
   ```
   shrinkResources true
   debuggable false
   minifyEnabled true
   proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
   ````
        
2. Distribute a new version with same config
3. When the dialog appears for update click download
4. Clicking download crashes the.

### **Details**

1. Which SDK version are you using?
    v 4.1.1
2. Which OS version did you experience the issue on?
    Android 6 to Android 10
3. What device version did you see this error on?  Were you using an emulator or a physical device?
    -Happening on all devices
4. What third party libraries are you using?
    - has no relation

Crash logs
```
Fatal Exception: java.lang.RuntimeException: An error occurred while executing doInBackground()
       at android.os.AsyncTask$3.done(AsyncTask.java:353)
       at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
       at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
       at java.util.concurrent.FutureTask.run(FutureTask.java:271)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
       at java.lang.Thread.run(Thread.java:764)
       
 Caused by java.lang.VerifyError: Verifier rejected class m.j.a.n.j.c.c: java.lang.Object m.j.a.n.j.c.c.doInBackground(java.lang.Object[]) failed to verify: java.lang.Object m.j.a.n.j.c.c.doInBackground(java.lang.Object[]): [0x6F] expected to be within a catch-all for an instruction where a monitor is held (declaration of 'm.j.a.n.j.c.c' appears in base.apk)
       at com.microsoft.appcenter.distribute.download.manager.DownloadManagerReleaseDownloader.request(DownloadManagerReleaseDownloader.java:121)
       at com.microsoft.appcenter.distribute.download.manager.DownloadManagerReleaseDownloader.onStart(DownloadManagerReleaseDownloader.java:141)
       at com.microsoft.appcenter.distribute.download.manager.DownloadManagerUpdateTask.doInBackground(DownloadManagerUpdateTask.java:35)
       at com.microsoft.appcenter.distribute.download.manager.DownloadManagerUpdateTask.doInBackground(DownloadManagerUpdateTask.java:20)
       at android.os.AsyncTask$2.call(AsyncTask.java:333)
       at java.util.concurrent.FutureTask.run(FutureTask.java:266)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
       at java.lang.Thread.run(Thread.java:764)
```

**Comment by AnastasiaKubova** [comment]
Hi, @build3r ! Thanks for getting in touch with us! I am trying to reproduce your issue by following steps from the description but for me all work well.
Could you please 
- Clarify does it happen before allowing installs from the unknown sources dialog?
- Share verbose logs?
- Share the crash report without obfuscated symbols?

**Comment by build3r** [comment]
One correction the crash is happening on library v 4.1.0 which in integrate in the old app which needs to be updated
Here are the logs after enabling verbose logging
Also one more correction, its happening on download click not on going background.

```
2021-03-31 18:32:15.145 D/AppCenter: checkPendingLogs(group_core) pendingLogCount=0 batchTimeInterval=3000
2021-03-31 18:32:15.145 D/AppCenter: App Center initialized.
2021-03-31 18:32:15.146 D/AppCenter: removeGroup(group_distribute)
2021-03-31 18:32:15.146 D/AppCenter: removeGroup(group_distribute/one)
2021-03-31 18:32:15.146 D/AppCenter: addGroup(group_distribute)
2021-03-31 18:32:15.146 D/Instrumentation: App onCreate release - 2.5.9 OS: 10
2021-03-31 18:32:15.146 D/AppCenter: Network 100 is available.
2021-03-31 18:32:15.147 D/AppCenter: Network has been connected.
2021-03-31 18:32:15.147 D/AppCenter: checkPendingLogs(group_distribute) pendingLogCount=0 batchTimeInterval=3000
2021-03-31 18:32:15.147 D/AppCenter: addGroup(group_distribute/one)
2021-03-31 18:32:15.147 D/AppCenter: checkPendingLogs(group_distribute/one) pendingLogCount=0 batchTimeInterval=3000
2021-03-31 18:32:15.147 D/AppCenterDistribute: Distribute workflow will be resumed on activity resume event.
2021-03-31 18:32:15.152 D/AppCenter: Storing a log to the Persistence database for log type startService with flags=1
2021-03-31 18:32:15.159 D/AppCenter: Stored a log to the Persistence database for log type startService with databaseId=3
2021-03-31 18:32:15.159 D/AppCenter: enqueue(group_core) pendingLogCount=1
2021-03-31 18:32:15.159 D/AppCenter: checkPendingLogs(group_core) pendingLogCount=1 batchTimeInterval=3000
2021-03-31 18:32:15.202 D/FA: Debug-level message logging enabled
2021-03-31 18:32:15.223 D/AppCenterDistribute: Resume distribute workflow...
2021-03-31 18:32:15.223 D/AppCenterDistribute: InstallerPackageName=null
2021-03-31 18:32:15.223 D/AppCenterDistribute: Show default update dialog.
2021-03-31 18:32:15.246 D/AppCenterDistribute: Get latest release details...
2021-03-31 18:32:15.247 D/AppCenterDistribute: Check if we need to report release installation..
2021-03-31 18:32:15.247 D/AppCenterDistribute: Current release was already reported, skip reporting.
2021-03-31 18:32:15.249 V/AppCenterDistribute: Calling https://api.appcenter.ms/v0.1/public/sdk/apps/****************************848be1e0/releases/latest?release_hash=7ebb62ffd77ce6a325ea026a676f242d5083fad5e491b95de714edec2fd8768b...
2021-03-31 18:32:15.249 V/AppCenterDistribute: Headers: {}
2021-03-31 18:32:16.503 D/AppCenterDistribute: Schedule download...
2021-03-31 18:32:16.503 W/nlife.karmalif: Class m.j.a.n.j.c.c failed lock verification and will run slower.
    
    --------- beginning of crash
2021-03-31 18:32:16.504 E/AndroidRuntime: FATAL EXCEPTION: AsyncTask #2
    Process: ***package_name***, PID: 23729
    java.lang.RuntimeException: An error occurred while executing doInBackground()
        at android.os.AsyncTask$4.done(AsyncTask.java:399)
        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
        at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
        at java.util.concurrent.FutureTask.run(FutureTask.java:271)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:919)
     Caused by: java.lang.VerifyError: Verifier rejected class m.j.a.n.j.c.c: java.lang.Object m.j.a.n.j.c.c.doInBackground(java.lang.Object[]) failed to verify: java.lang.Object m.j.a.n.j.c.c.doInBackground(java.lang.Object[]): [0x6F] expected to be within a catch-all for an instruction where a monitor is held (declaration of 'm.j.a.n.j.c.c' appears in base.apk)
        at m.j.a.n.j.c.a.f(SourceFile:6)
        at m.j.a.n.j.c.d.doInBackground(SourceFile:7)
        at android.os.AsyncTask$3.call(AsyncTask.java:378)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:919) 
2021-03-31 18:32:16.506 D/AppCenter: Cleared pending log states
2021-03-31 18:32:16.506 D/AppCenter: Channel completed shutdown
```

**Comment by build3r** [comment]
happening on non-debuggable, minified + shrink resource builds on library v 4.1.1

**Comment by build3r** [comment]
> @build3r, Thanks for sharing the additional information. Just for clarification - this is happening only in release mode, in debug mode, all work well, right?
> Have you a chance to share the crash report without obfuscated symbols?

Yes @AnastasiaKubova Its happening only in release mode with follwing config
```
shrinkResources true
debuggable false
minifyEnabled true
shrinkResources true
debuggable false
minifyEnabled true
proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
```
>Have you a chance to share the crash report without obfuscated symbols?
looks like the obfuscated class are internals classes
de-obfuscation does nothing
![image](https://user-images.githubusercontent.com/6481548/113157071-304c0d80-9258-11eb-979b-e9627bd309f3.png)

**Comment by AnastasiaKubova** [comment]
Thanks for the additional information. I need some time to check it.

**Comment by build3r** [comment]
@AnastasiaKubova were you able to reproduce the issue?
Also is there any way we can make the older app version affected by this issue upgrade to new one?

**Comment by AnastasiaKubova** [comment]
Hi, @build3r ! I am trying to reproduce your issue, but for me still all work well. Could you please share some more details:
- Does this happen only with the mandatory update or with the usual update too?
- Does this happen before allowing installs from the unknown sources dialog or when these settings already setup?
- Could you please confirm that the download manager application is enabled?

**Comment by build3r** [comment]
@AnastasiaKubova 
I have created a minimal POC to reproduce the crash
you can find the repo here https://github.com/build3r/AppcenterCrash/tree/master
also see the video below:
https://github.com/build3r/AppcenterCrash/blob/master/video.mp4


> Does this happen only with the mandatory update or with the usual update too?

Happens on both

>Does this happen before allowing installs from the unknown sources dialog or when these settings already setup?  

Happens after the  unknown sources dialog 

>Could you please confirm that the download manager application is enabled?  

Yes it is enabled and the updates work on sdk v 4.1.0

**Comment by AnastasiaKubova** [comment]
Thanks for sharing a sample project. I will check it and notify you about the results.

**Comment by build3r** [comment]
Was able to narrow down the issue to this config
```
debuggable false
minifyEnabled true
proguardFiles getDefaultProguardFile('proguard-android-optimize.txt')
```
If we remove the proguard-android-optimize.txt option the issue is not observed

**Comment by build3r** [comment]
@AnastasiaKubova 
This is the un-obfuscated stack trace
```
    java.lang.RuntimeException: An error occurred while executing doInBackground()
        at android.os.AsyncTask$4.done(AsyncTask.java:399)
        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
        at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
        at java.util.concurrent.FutureTask.run(FutureTask.java:271)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:919)
     Caused by: java.lang.VerifyError: Verifier rejected class com.microsoft.appcenter.distribute.download.manager.DownloadManagerRequestTask: java.lang.Void com.microsoft.appcenter.distribute.download.manager.DownloadManagerRequestTask.a() failed to verify: java.lang.Void com.microsoft.appcenter.distribute.download.manager.DownloadManagerRequestTask.a(): [0x6C] expected to be within a catch-all for an instruction where a monitor is held (declaration of 'com.microsoft.appcenter.distribute.download.manager.DownloadManagerRequestTask' appears in base.apk)
        at com.microsoft.appcenter.distribute.download.manager.DownloadManagerReleaseDownloader.f(:2)
        at com.microsoft.appcenter.distribute.download.manager.DownloadManagerUpdateTask.a(:3)
        at com.microsoft.appcenter.distribute.download.manager.DownloadManagerUpdateTask.doInBackground(Unknown Source:2)
        at android.os.AsyncTask$3.call(AsyncTask.java:378)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:919) 
```

**Comment by build3r** [comment]
Also the following proguard rules fixes the issue
```
-keepclasseswithmembers class * extends com.microsoft.appcenter.distribute.download.ReleaseDownloader {
    <fields>;
    <methods>;
}
```

But is there any way to upgrade the users having the above issue?
There are >1600 users in our public beta version who are currently not able to update because of this issue.

**Comment by AnastasiaKubova** [comment]
Hi, @build3r ! Thanks for sharing the sample project. I was able to reproduce your issue. I will label this issue as a bug. 
Also, I recommend you remove the `app secret` value from your test application. See more about the [protection of app-secret](https://docs.microsoft.com/en-us/appcenter/sdk/troubleshooting/android#protect-the-app-center-secret-value) value.

> But is there any way to upgrade the users having the above issue?

As an alternative way, you can try to download the update of your application via the [install portal](https://docs.microsoft.com/en-us/appcenter/distribution/testers/testing-android#installing-an-app-via-install-portal).

**Comment by build3r** [comment]
Thanks @AnastasiaKubova 
>I recommend you remove the app secret value from your test application. See more about the protection of app-secret value.

Have removed, but anyways I created that project only for reproducing the bug, Will delete it now.

**Comment by AnastasiaKubova** [comment]
The fix has been released in the [4.2.0](https://github.com/microsoft/appcenter-sdk-android/releases/tag/4.2.0) version of App Center SDK. Please update the App Center SDK to the latest version and let me know if this issue still persists.
